akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -436,23 +447,56 @@ final class CoordinatedShutdown private[akka] (
    * concurrently, but they are running before Akka internal shutdown
    * hooks, e.g. those shutting down Artery.
    */
-  @tailrec def addJvmShutdownHook[T](hook: ⇒ T): Unit = {
+  def addJvmShutdownHook[T](hook: ⇒ T): Unit = addCancellableJvmShutdownHook(hook)
+
+  /**
+   * Scala API: Add a JVM shutdown hook that will be run when the JVM process
+   * begins its shutdown sequence. Added hooks may run in any order
+   * concurrently, but they are running before Akka internal shutdown
+   * hooks, e.g. those shutting down Artery.
+   *
+   * The returned ``Cancellable`` makes it possible to de-register the hook. For example
+   * on actor system shutdown to avoid leaking references to the actor system in tests.
+   *
+   * For shutdown hooks that does not have any requirements on running before the Akka
+   * shutdown hooks the standard library JVM shutdown hooks APIs are better suited.
+   */
+  @tailrec def addCancellableJvmShutdownHook[T](hook: ⇒ T): Cancellable = {
     if (!runStarted.get) {
       val currentLatch = _jvmHooksLatch.get
       val newLatch = new CountDownLatch(currentLatch.getCount.toInt + 1)
       if (_jvmHooksLatch.compareAndSet(currentLatch, newLatch)) {
-        try Runtime.getRuntime.addShutdownHook(new Thread {
+        val thread = new Thread {
           override def run(): Unit = {
             try hook finally _jvmHooksLatch.get.countDown()
           }
-        }) catch {
+        }
+        thread.setName(s"${system.name}-shutdown-hook-${newLatch.getCount}")
+        try {
+          Runtime.getRuntime.addShutdownHook(thread)
+          new Cancellable {
+            @volatile var cancelled = false
+            def cancel(): Boolean = {
+              if (Runtime.getRuntime.removeShutdownHook(thread)) {
```

Comment:
this will throw `IllegalStateException` if called when shutdown is in progress, should be propagate that or discard?